text/0000-grid-primitives.md

```diff
@@ -0,0 +1,761 @@
+It is left to the smart-contract implementer to determine if this would result
+in an error condition.
+
+### Struct
+
+A struct value would be represented as follows:
+
+```
+PropertyDefinition(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_properties=[
+        PropertyDefinition(
+            name='speed',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+        PropertyDefinition(
+            name='duration',
+            data_type=PropertyDefinition.DataType.NUMBER,
+            number_exponent=-6),
+    ],
+    required=True
+)
+```
+
+An instance of the “shock" struct would be as follows:
+
+```
+PropertyValue(
+    name='shock',
+    data_type=PropertyDefinition.DataType.STRUCT,
+    struct_values=[
+        PropertyValue(
+            name='speed',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=500000),
+        PropertyValue(
+            name='duration',
+            data_type=PropertySchema.DataType.NUMBER,
+            number_value=10000)
+        ])
+```
+
+The property value for a struct must contain all the struct values from the
+property definition, or it is invalid.  The defaults for the struct values
+themselves depend on their data types and/or the smart-contract implementer
+validation rules.
+
+### Latitude/Longitude
+
+A latitude/longitude (lat/long) value would be represented as follows:
+
+```
+PropertyDefintion(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    required=True
+)
+```
+
+A lat/long instance would be as follows:
+
+```
+PropertyValue(
+    name='origin',
+    data_type=PropertyDefinition.DataType.LAT_LONG,
+    lat_long_value=LatLong(
+        latitude=44977753,
+        longitude=-93265015)
+)
+```
+
+Due to the use of protobuf, the default values for `LatLong` would be `(0, 0)`.
+While this is a valid lat/long, it could be used to indicate an error, depending
+on the choice of the smart-contract implementer.
+
+### Datetime
+
+A datetime value would be represented as follows:
+```
+PropertyDefinition(
+    name='created_at',
+    data_type=PropertyDefinition.DataType.DATETIME,
+    required=True
+)
+```
+
+A datetime instance for this definition would be as follows:
+```
+PropertyValue(
+    name='created_at',
+    data_type=PropertyDefinition.DataType.DATETIME,
+    datetime_value='2019-05-31T14:53:18+0000'
+)
+```
+
+Due to the use of protobuf, the default value for `datetime_value` is an empty
+string, and therefore an invalid value.  This field must be set for the
+`DATETIME` data type.
+
+## Schema Example
+
+A complete object representation can be built from the property definition
+messages, and instances can be represented by constructing items with the
+property value messages.  
+
+Suppose there is a requirement to store different types of light bulbs as part
+of an application. A lightbulb may consist of the following properties: size,
```

Comment:
s/lightbulb/light bulb/

Reply:
That usage is consistent with the rest of the document.

Reply:
Like Logan, my personal preference is for "light bulb", but "lightbulb" is acceptable, according to Merriam-Webster as far back as 2009. (Yes, I looked it up.) Plus, consistency is good.

Reply:
So did I - that's how I arrived at that construction - it was more to inform if Lightbulb (the type) should be camel-cased or not. Every other usage followed.